### PR TITLE
Show Disentanglement

### PR DESCRIPTION
**Pull request by BartMelman**
I changed the function show_disentanglement such that the latent traversals, heat maps and reconstructions are combined to create a single image. There is the possibility to add the KL divergence to the right. The heat maps are colored.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -13,7 +13,6 @@
                            get_config_section, update_namespace_, FormatterNoDuplicate)
 from viz.visualize import Visualizer
 
-
```
Comment (alecokas) [inline main.py]: This change is not needed. Add the line back :stuck_out_tongue:

File: main_concatenate.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+# def combine_images(width, height,image_list, nr_columns,):
+#     new_image = PIL.Image.new("RGB", (max(h1, h2), w1+w2))
+
+def concatenate_images(image_list, width, height, nr_columns = 1, percentage_space = 0.0):
+    total_nr_images = len(image_list)  
+    nr_rows = int(ceil(float(total_nr_images)/nr_columns))
+    # image_size = image_list[0].size
+    # width = image_size[0]
+    # height = image_size[1]
```
Comment (alecokas) [inline main_concatenate.py]: Remove these comments

File: main_concatenate.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
+import PIL
+from PIL import Image
+import os
+from math import ceil, floor
+
+IMGS_DIR = "imgs"
+
+# def combine_images(width, height,image_list, nr_columns,):
+#     new_image = PIL.Image.new("RGB", (max(h1, h2), w1+w2))
```
Comment (alecokas) [inline main_concatenate.py]: Please remove these comments since they aren't needed instructive comments.

File: main_concatenate.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    image_list = get_image_list(image_file_name_list)
+    width_max, height_max = get_max_size(image_list)
+    percentage_space  = 0.03
+    new_image = concatenate_images(image_list, width_max, height_max, nr_columns, percentage_space)
+    
+    new_image.save(os.path.join(IMGS_DIR, output_file_name + '.png'))
+    return True
+
+if __name__ == '__main__':
+    main()
```
Comment (alecokas) [inline main_concatenate.py]: Add Docstring comments to these functions as in the main code for readability.

File: viz/log_plotter.py [hunk 5]
```diff
@@ -31,4 +31,4 @@ def plot(self, log_dir):
         x_axis = list(df)[0]
         df.plot(x=x_axis, y=headers, grid=True, style='-')
 
-        return plt
+        return plt
```
Comment (alecokas) [inline viz/log_plotter.py]: Please revert this change. It is not doing anything.

File: viz/visualize.py [hunk 6]
```diff
@@ -49,122 +54,68 @@ def __init__(self, model, dataset, model_dir=None, save_images=True,
         self.dataset = dataset
         self.loss_of_interest = loss_of_interest
 
+    def tensor_gray_scale_to_color(self, input_tensor):
+        # input_tensor, consists of gray_scale values and has dimension: latent dim, gray_scale value, x_position, y_position
```
Comment (alecokas) [inline viz/visualize.py]: Add docstrings instead of this comment

File: viz/visualize.py [hunk 7]
```diff
@@ -49,122 +54,68 @@ def __init__(self, model, dataset, model_dir=None, save_images=True,
...
+                    output[latent_dim, 2, x_posn, y_posn] = min_color[2] + (max_color[2]-min_color[2])*scale_id
+        return output
+
     def show_disentanglement_fig2(self, reconstruction_data, latent_sweep_data, heat_map_data,
-                                  latent_order=None, heat_map_size=(32, 32), filename='show-disentanglement.png',
-                                  size=8, sample_latent_space=None):
+                                  latent_order=None, heat_map_size=(32, 32), file_name='show-disentanglement.png',
+                                  size=8, sample_latent_space=None, base_directory = '',select_prior=False,show_text=False):
         """ Reproduce Figure 2 from Burgess https://arxiv.org/pdf/1804.03599.pdf
             TODO: STILL TO BE IMPLEMENTED
```
Comment (alecokas) [inline viz/visualize.py]: Remove the TODO (since you did do it) and add a proper Docstring

File: viz/visualize.py [hunk 8]
```diff
@@ -49,122 +54,68 @@ def __init__(self, model, dataset, model_dir=None, save_images=True,
...
-        new_torch = torch.tensor(np.zeros((self.model.latent_dim, 1, 64, 64)))
-        for i in range(0, self.model.latent_dim):
-            new_torch[i, 0, :, :] = heat_map_sorted[i]
-        nr_imgs_per_latent = size
+        image_file_name_list = [
+            os.path.join(base_directory, 'recon_comp.png'),
+            os.path.join(base_directory, 'heatmap.png'),
+            ''
+            ]
+        # === get reconstruction === #
```
Comment (alecokas) [inline viz/visualize.py]: This ` ====== ` is not a nice comment style

File: viz/visualize.py [hunk 9]
```diff
@@ -484,20 +453,19 @@ def prior_traversal(self, sample_latent_space=None, reorder_latent_dims=False, n
...
-            )
+                )       
+
 
     def all_latent_traversals(self, sample_latent_space=None, size=8):
         """
         Traverses all latent dimensions one by one and plots a grid of images
         where each row corresponds to a latent traversal of one latent
         dimension.
-
```
Comment (alecokas) [inline viz/visualize.py]: These spaces should not be deleted

File: viz/viz_helpers.py [hunk 10]
```diff
@@ -116,21 +115,28 @@ def add_labels(label_name, tensor, num_rows, sorted_list, dataset):
...
     return traversal_images_with_text   
 
 
-def upsample(input_data, scale_factor):
+def upsample(input_data, scale_factor, colour_flag=False):
+
     """ TODO: Bart add Docstring - I don't really know what this is
     """
     # duplicate
+
```
Comment (alecokas) [inline viz/viz_helpers.py]: Attend to this Docstring

File: viz/visualize.py [hunk 11]
```diff
@@ -49,122 +54,68 @@ def __init__(self, model, dataset, model_dir=None, save_images=True,
...
-        ]
-
-        new_torch = torch.tensor(np.zeros((self.model.latent_dim, 1, 64, 64)))
-        for i in range(0, self.model.latent_dim):
-            new_torch[i, 0, :, :] = heat_map_sorted[i]
-        nr_imgs_per_latent = size
+        image_file_name_list = [
+            os.path.join(base_directory, 'recon_comp.png'),
+            os.path.join(base_directory, 'heatmap.png'),
+            ''
```
Comment (alecokas) [inline viz/visualize.py]: What is the purpose of this empty string? Why is it required?

File: viz/visualize.py [hunk 12]
```diff
@@ -484,20 +453,19 @@ def prior_traversal(self, sample_latent_space=None, reorder_latent_dims=False, n
                 filename=file_name,
                 nrow=num_increments,
                 pad_value=(1 - get_background(self.dataset))
-            )
+                )
```
Comment (alecokas) [inline viz/visualize.py]: Not sure why this change was made

File: viz/viz_helpers.py [hunk 13]
```diff
@@ -153,3 +159,9 @@ def make_grid_img(tensor, **kwargs):
     img_grid = grid.mul_(255).add_(0.5).clamp_(0, 255).permute(1, 2, 0)
     img_grid = img_grid.to('cpu', torch.uint8).numpy()
     return img_grid
+
+def get_image_list(image_file_name_list):
+    image_list = []
+    for file_name in image_file_name_list:
+        image_list.append(Image.open(file_name))
+    return image_list
```
Comment (alecokas) [inline viz/viz_helpers.py]: There should be a newline at the end of this file.

File: main_concatenate.py [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+def concatenate_images(image_list, width, height, nr_columns = 1, percentage_space = 0.0):
+    total_nr_images = len(image_list)  
+    nr_rows = int(ceil(float(total_nr_images)/nr_columns))
+    # image_size = image_list[0].size
+    # width = image_size[0]
+    # height = image_size[1]
+    space = int(round(percentage_space*max(width, height)))
+    new_image = PIL.Image.new("RGB", (width*nr_columns, height*nr_rows))
+    print(width,height,nr_columns,nr_rows)
+    # === create new image === #
```
Comment (alecokas) [inline main_concatenate.py]: This is not a nice style for comments and I am sure it violates PEP8 in some way.

File: main_concatenate.py [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+def concatenate_images(image_list, width, height, nr_columns = 1, percentage_space = 0.0):
+    total_nr_images = len(image_list)  
+    nr_rows = int(ceil(float(total_nr_images)/nr_columns))
+    # image_size = image_list[0].size
+    # width = image_size[0]
+    # height = image_size[1]
+    space = int(round(percentage_space*max(width, height)))
+    new_image = PIL.Image.new("RGB", (width*nr_columns, height*nr_rows))
+    print(width,height,nr_columns,nr_rows)
```
Comment (alecokas) [inline main_concatenate.py]: Remove this `print()`

File: main_concatenate.py [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+    # === create new image === # 
+    for id_image in range(0, len(image_list)):
```
Comment (alecokas) [inline main_concatenate.py]: No need to explicitly start the range from 0

File: main_concatenate.py [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+
+    Parameters
+    ----------
+    args: argparse.Namespace
+        Arguments
+    """
+    image_file_name_list = [
+    'show_disentanglement.png',
+    'show_disentanglement.png',
+    'show_disentanglement.png'
```
Comment (alecokas) [inline main_concatenate.py]: Why is the same thing repeated 3 times here? Is there no way to choose the input and output files from the command line?

File: viz/visualize.py [hunk 18]
```diff
@@ -197,28 +148,43 @@ def generate_heat_maps(self, data, reorder=False, heat_map_size=(32, 32), file_n
...
+                heat_map_list = [heat_map_color[i,:,:,:] for i in range(num_latent_dims)]
 
                 loss_list = read_loss_from_file(os.path.join(self.model_dir, TRAIN_FILE), loss_to_fetch=self.loss_of_interest)
-                heat_map = self.reorder(list_to_reorder=heat_map_list, reorder_by_list=loss_list)
-                heat_map = heat_map[:, None, :, :]
-
+                heat_map_color = self.reorder(list_to_reorder=heat_map_list, reorder_by_list=loss_list)
             # Normalise between 0 and 1
-            heat_map = (heat_map - torch.min(heat_map)) / (torch.max(heat_map) - torch.min(heat_map))
+            # heat_map = (heat_map - torch.min(heat_map)) / (torch.max(heat_map) - torch.min(heat_map))
```
Comment (alecokas) [inline viz/visualize.py]: Remove comment

File: viz/visualize.py [hunk 19]
```diff
@@ -11,13 +11,18 @@
 from viz.viz_helpers import (reorder_img, read_loss_from_file, add_labels,
                              upsample, make_grid_img)
 
+import PIL
+from PIL import Image
```
Comment (alecokas) [inline viz/visualize.py]: Either `import PIL` or import the exact things you need from PIL, but don't import `Image` twice

File: viz/visualize.py [hunk 20]
```diff
@@ -11,13 +11,18 @@
...
+import os
+from math import ceil, floor
```
Comment (alecokas) [inline viz/visualize.py]: Put these imports up with the other ones which are similar to them

File: main_concatenate.py (hunk 4 above)
Comment (BartMelman) [reply main_concatenate.py]: What is a docstring? And how do I do that?

File: main_concatenate.py (hunk 17 above)
Comment (BartMelman) [reply main_concatenate.py]: It is going to be a long list and I do not know how all the files are going to be called.

File: main_concatenate.py (hunk 4 above)
Comment (alecokas) [reply main_concatenate.py]: I see you have done it.

**Comment by alecokas** [comment]
I'm just going to merge this and make the changes on my side for convenience. They are all formatting changes that are required.

File: viz/viz_helpers.py (hunk 10 above)
Comment (BartMelman) [reply viz/viz_helpers.py]: Done

File: viz/visualize.py (hunk 11 above)
Comment (BartMelman) [reply viz/visualize.py]: adapted

File: viz/visualize.py (hunk 12 above)
Comment (BartMelman) [reply viz/visualize.py]: Reverted
